File: test/cpp/hello.cpp

#include <cstdint>
#include <iostream>
#include <string>

#define ryo_trace(message) (::std::cerr << (message) << " (" << __FILE__ << "#" << __LINE__ << ")" << ::std::endl)

using Result = int;

struct Point;

Result main();
::std::int32_t after(const ::std::int32_t x);
float norm2(const Point& point);

struct Point {
  float x = float(0);
  float y = float(0);
};

Result main() {
  const auto x = ::std::int32_t(3);
  const auto y = float(-3.5);
  ryo_trace("Hello!");
  ryo_trace(after(x));
  ryo_trace(after(::std::int32_t(y)));
  const Point point = {float(4), y};
  ryo_trace(norm2(point));
  return int(0);
}

::std::int32_t after(const ::std::int32_t x) {
  return ::std::int32_t.add(x, ::std::int32_t(1));
}

float norm2(const Point& point) {
  return float.add(float.mul(point.x, point.x), float.mul(point.y, point.y));
}

File: test/hello.ts
import {f32, i32, int, trace} from 'ryo';

type Result = int;

class Point {
  x: f32 = 0;
  y: f32 = 0;
}

function main(): Result {
  let x: i32 = 3;
  let y = -3.5;
  trace('Hello!');
  trace(after(x));
  trace(after(y));
  let point: Point = {x: 4, y};
  // Way to embed an object literal as an expression with out of order properties.
  // Of course, if we are sure about no side effects, we can reorder operations.
  // let point: Point = {y: something(), x: other()};
  // const Point point = ([](){double y = something(); double x = other(); return Point {x, y};})();
  // const Point point = ([](){double y = something(); return Point {other(), y};})();
  trace(norm2(point));
  return 0;
}

function after(x: i32): i32 {
  return i32(x + 1);
}

function norm2(point: Point): number {
  return point.x * point.x + point.y * point.y;
}

main();
